### PR TITLE
Issue 571: Added Maximum Attending Limit at the event level.

### LINKED ISSUE #571: Maximum Attending Limit at the event level
Maximum Attending Limit is a global setting that applies to all events. Let's create this setting at the event level as well. It will inherit from that global setting, but you can change it per event.

### PR DESCRIPTION
**Pull request by bluecollarcoders**
### Description of the Change

Benefits realized: Maximum Attending Limit is now on the event level

Screen Shot 2024-03-01 at 2 45 03 AM (2)

Closes #571 

### How to test the Change

In the event setting you will see a panel body labeled Maximum Attending Limit. Set the limit with the number control.

### Changelog Entry

> Added - New feature

### Credits

Props 
@bluecollarcoders 

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/GatherPress/gatherpress/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

File: includes/core/classes/class-assets.php [hunk 1]
```diff
@@ -297,6 +297,7 @@ protected function localize( int $post_id ): array {
 				'currentUser'         => $event->rsvp->get( get_current_user_id() ),
 				'dateTime'            => $event->get_datetime(),
 				'enableAnonymousRsvp' => (bool) get_post_meta( $post_id, 'enable_anonymous_rsvp', true ),
+				'maxAttendance'       => (int) get_post_meta( $post_id, 'max_attendance', true ),
```
Comment (mauteri) [inline includes/core/classes/class-assets.php:300]: Let's change this everywhere to `maxAttendanceLimit` and `max_attendance_limit`. Looks good though!

**Review by mauteri — CHANGES_REQUESTED** [review]

File: src/components/MaxAttendance.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @return {JSX.Element} A number control for setting the maximum attendance limit.
+ */
+const MaxAttendance = () => {
+	const { editPost, unlockPostSaving } = useDispatch('core/editor');
+	const isNewEvent = useSelect((select) => {
+		return select('core/editor').isCleanNewPost();
+	}, []);
+
+	// eslint-disable-next-line no-shadow
+	let defaultMaxAttendance = useSelect((select) => {
```
Comment (mauteri) [inline src/components/MaxAttendance.js:32]: `defaultMaxAttendanceLimit`

File: src/components/MaxAttendance.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		return select('core/editor').getEditedPostAttribute('meta')
+			.max_attendance;
+	}, []);
+
+	if (isNewEvent) {
+		defaultMaxAttendance = getFromGlobal('settings.maxAttendance');
+	}
+
+	const [maxAttendance, setMaxAttendance] = useState(null);
```
Comment (mauteri) [inline src/components/MaxAttendance.js:41]:
```
const [maxAttendanceLimit, setMaxAttendanceLimit] = useState(defaultMaxAttendanceLimit);
```

**Review by mauteri — CHANGES_REQUESTED** [review]

**Comment by bluecollarcoders** [comment]
@mauteri I updated the PR, but it says there are conflicts with the branch. I did a git pull and I had no conflicts with the branch.

**Comment by mauteri** [comment]
@bluecollarcoders make sure on your forked repo you sync main with your main then pull that down to your branch. you'll also need to run `npm install` with the latest version of node in `.npmrc` and `npm run build`. Let me know if you have any other questions. Thx!

File: includes/core/classes/class-assets.php [hunk 4]
```diff
@@ -297,14 +297,14 @@ protected function localize( int $post_id ): array {
 
 		if ( ! empty( $event->event ) ) {
 			$event_details = array(
-				'currentUser'          => $event->rsvp->get( get_current_user_id() ),
-				'dateTime'             => $event->get_datetime(),
-				'enableAnonymousRsvp'  => (bool) get_post_meta( $post_id, 'enable_anonymous_rsvp', true ),
-				'enableInitialDecline' => (bool) get_post_meta( $post_id, 'enable_initial_decline', true ),
```
Comment (mauteri) [inline includes/core/classes/class-assets.php:303]: This one shouldn't have been removed.

File: includes/core/classes/class-assets.php [hunk 5]
```diff
@@ -318,12 +318,12 @@ protected function localize( int $post_id ): array {
 				'unregisterBlocks' => $this->unregister_blocks(),
 			),
 			'settings'     => array(
-				'dateFormat'           => $settings->get_value( 'general', 'formatting', 'date_format' ),
-				'enableAnonymousRsvp'  => ( 1 === (int) $settings->get_value( 'general', 'general', 'enable_anonymous_rsvp' ) ),
-				'enableInitialDecline' => ( 1 === (int) $settings->get_value( 'general', 'general', 'enable_initial_decline' ) ),
```
Comment (mauteri) [inline includes/core/classes/class-assets.php:323]: same with this one.

File: src/components/MaxAttendanceLimit.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+const MaxAttendanceLimit = () => {
+	const { editPost, unlockPostSaving } = useDispatch('core/editor');
+	const isNewEvent = useSelect((select) => {
+		return select('core/editor').isCleanNewPost();
+	}, []);
+
+	// eslint-disable-next-line no-shadow
+	let defaultMaxAttendanceLimit = useSelect((select) => {
+		return select('core/editor').getEditedPostAttribute('meta')
+			.max_attendance;
```
Comment (mauteri) [inline src/components/MaxAttendanceLimit.js:34]: this should be `.max_attendance_limit`.

File: src/components/MaxAttendanceLimit.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	}, []);
+
+	if (isNewEvent) {
+		defaultMaxAttendanceLimit = getFromGlobal('settings.maxAttendance');
```
Comment (mauteri) [inline src/components/MaxAttendanceLimit.js:38]: `defaultMaxAttendanceLimit = getFromGlobal('settings.maxAttendanceLimit');`

File: src/components/MaxAttendanceLimit.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		[editPost, unlockPostSaving]
+	);
+
+	useEffect(() => {
+		setMaxAttendanceLimit(defaultMaxAttendanceLimit);
+	}, [defaultMaxAttendanceLimit]);
+
+	return (
+		<NumberControl
+			label={__('Maximum Attending Limit', 'gatherpress')}
```
Comment (mauteri) [inline src/components/MaxAttendanceLimit.js:62]: `Maximum Attendance Limit` for consistency.

**Review by mauteri — CHANGES_REQUESTED** [review]
Just a few small things and I think this will be good. Thx @bluecollarcoders!

File: includes/core/classes/class-assets.php [hunk 9]
```diff
@@ -301,6 +301,7 @@ protected function localize( int $post_id ): array {
 				'dateTime'             => $event->get_datetime(),
 				'enableAnonymousRsvp'  => (bool) get_post_meta( $post_id, 'enable_anonymous_rsvp', true ),
 				'enableInitialDecline' => (bool) get_post_meta( $post_id, 'enable_initial_decline', true ),
+				'maxAttendanceLimit'   => (int) get_post_meta( $post_id, 'max_attendance_limit', true ),
```
Comment (mauteri) [inline includes/core/classes/class-assets.php:304]: My PR merged, so let's update this.
```
'maxAttendanceLimit'   => (int) get_post_meta( $post_id, 'gatherpress_max_attendance_limit', true ),
```

File: src/components/MaxAttendanceLimit.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+const MaxAttendanceLimit = () => {
+	const { editPost, unlockPostSaving } = useDispatch('core/editor');
+	const isNewEvent = useSelect((select) => {
+		return select('core/editor').isCleanNewPost();
+	}, []);
+
+	// eslint-disable-next-line no-shadow
+	let defaultMaxAttendanceLimit = useSelect((select) => {
+		return select('core/editor').getEditedPostAttribute('meta')
+			.max_attendance_limit;
```
Comment (mauteri) [inline src/components/MaxAttendanceLimit.js:34]:
```
.gatherpress_max_attendance_limit;
```

**Review by mauteri — CHANGES_REQUESTED** [review]

**Comment by mauteri** [comment]
Merging this, thx @bluecollarcoders! I opened a ticket with `wordpress-version-checker` on this issue https://github.com/skaut/wordpress-version-checker/issues/1193 (cc @carstingaxion @jeffpaul)

**Pull request merged**